ENH: Fast paths for richcompare

ganesh-k13 opened the pull request:
## Changes:
1. Added check to see if RHS is a python number and prevent the creation of array.
2. Added bench

## Benchmarks:
<details>
<summary>Benchmarks:</summary>

<pre>                                                                                                                   
********************************************************************************                                                                          
WARNING: you have uncommitted changes --- these will NOT be benchmarked!                                                                                  
********************************************************************************                                                                          
· Creating environments                                                      
· Discovering benchmarks                                                     
·· Uninstalling from virtualenv-py3.8-Cython                                 
·· Building 933620b4 <enh_14415_fast_compare> for virtualenv-py3.8-Cython..................................                                                                                                                                                                                                          
·· Installing 933620b4 <enh_14415_fast_compare> into virtualenv-py3.8-Cython.                                                                             
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)                                                                                  
[  0.00%] · For numpy commit 4d290795 <master> (round 1/2):                                                                                               
[  0.00%] ·· Building for virtualenv-py3.8-Cython.                                                                                                                                                                                                                                                                   
[  0.00%] ·· Benchmarking virtualenv-py3.8-Cython                                                                                                         
[ 25.00%] ··· Running (bench_ufunc.CustomScalarEqual.time_scalar_equality--).                                                                             
[ 25.00%] · For numpy commit 933620b4 <enh_14415_fast_compare> (round 1/2):                                                                               
[ 25.00%] ·· Building for virtualenv-py3.8-Cython.                                                                                                        
[ 25.00%] ·· Benchmarking virtualenv-py3.8-Cython                                                                                                         
[ 50.00%] ··· Running (bench_ufunc.CustomScalarEqual.time_scalar_equality--).                                                                             
[ 50.00%] · For numpy commit 933620b4 <enh_14415_fast_compare> (round 2/2):                                                                               
[ 50.00%] ·· Benchmarking virtualenv-py3.8-Cython                                                                                                         
[ 75.00%] ··· bench_ufunc.CustomScalarEqual.time_scalar_equality                                                                                                              ok                                                                                                                                     
[ 75.00%] ··· =============== ==========                                     
                   dtype                                                                                                                                  
              --------------- ----------                                     
                 numpy.int8    439±6ms                                       
                numpy.int16    445±8ms                                       
                numpy.int32    435±10ms                                      
                numpy.int64    147±2ms                                       
                numpy.uint8    427±10ms                                      
                numpy.uint16   444±8ms                                       
                numpy.uint32   437±10ms                                      
                numpy.uint64   447±4ms                                       
               numpy.float32   471±10ms                                      
               numpy.float64   180±4ms                                                                                                                    
              =============== ==========                                                                                                                  
                                                                             
[ 75.00%] · For numpy commit 4d290795 <master> (round 2/2):                                                                                               
[ 75.00%] ·· Building for virtualenv-py3.8-Cython.                                                                                                        
[ 75.00%] ·· Benchmarking virtualenv-py3.8-Cython                                                                                                         
[100.00%] ··· bench_ufunc.CustomScalarEqual.time_scalar_equality                                                                                                              ok                                                                                                                                     
[100.00%] ··· =============== ============                                   
                   dtype                                                     
              --------------- ------------                                   
                 numpy.int8     1.35±0s                                      
                numpy.int16    1.34±0.01s                                    
                numpy.int32    1.34±0.02s                                    
                numpy.int64     147±3ms                                      
                numpy.uint8     1.34±0s                                      
                numpy.uint16    1.36±0s                                      
                numpy.uint32    1.35±0s                                      
                numpy.uint64   1.54±0.01s                                    
               numpy.float32    1.54±0s                                      
               numpy.float64    184±4ms                                      
              =============== ============                                   

       before           after         ratio                                  
     [4d290795]       [933620b4]                                             
     <master>         <enh_14415_fast_compare>                                                                                                            
-      1.34±0.01s          445±8ms     0.33  bench_ufunc.CustomScalarEqual.time_scalar_equality(<class 'numpy.int16'>)                                                                                                                                                                                               
-         1.36±0s          444±8ms     0.33  bench_ufunc.CustomScalarEqual.time_scalar_equality(<class 'numpy.uint16'>)                                                                                                                                                                                              
-      1.34±0.02s         435±10ms     0.33  bench_ufunc.CustomScalarEqual.time_scalar_equality(<class 'numpy.int32'>)                                                                                                                                                                                               
-         1.35±0s          439±6ms     0.33  bench_ufunc.CustomScalarEqual.time_scalar_equality(<class 'numpy.int8'>)                                                                                                                                                                                                
-         1.35±0s         437±10ms     0.32  bench_ufunc.CustomScalarEqual.time_scalar_equality(<class 'numpy.uint32'>)                                                                                                                                                                                              
-         1.34±0s         427±10ms     0.32  bench_ufunc.CustomScalarEqual.time_scalar_equality(<class 'numpy.uint8'>)                                                                                                                                                                                               
-         1.54±0s         471±10ms     0.31  bench_ufunc.CustomScalarEqual.time_scalar_equality(<class 'numpy.float32'>)                                                                                                                                                                                             
-      1.54±0.01s          447±4ms     0.29  bench_ufunc.CustomScalarEqual.time_scalar_equality(<class 'numpy.uint64'>)                                                                                                                                                                                              

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.                                  
PERFORMANCE INCREASED.                                                       

</pre>
</details>

resolves: #14415

cc: @seberg , @eric-wieser , @mattip

ganesh-k13 committed:
BENCH: Added benchmarks for fast paths (#17970)

seberg commented:
Thanks, I will have to think about it more. I do guess that it might be nice to intercept some of this even earlier, but I do not mind injecting a fast-path there it is practical and doesn't add much complexity.

One thing to think about `.item()` has two awkward cases:

1. Low precision floating point numbers are upcast to float64, I expect that is fine, but did not think it through.
2. High precision floating point numbers (longdouble, clongdouble) are *unchanged*, that means the fast-path is definitely broken for them. Maybe that is the test-failure, maybe not.

(To be fair, the 2. part is already broken in some similar paths, but this would break it more/worse.)

ganesh-k13 commented:
Thanks for the review @seberg , will test it out :+1:

ganesh-k13 commented:
I tried something like this:
```c
* #Name = Long*11, Float*3,,,#
* #fastpath = 1*14, 0*3#

.....

   case -2:
        /* use ufunc */
        if (PyErr_Occurred()) {
            return NULL;
        }
#if @fastpath@
        if (PyArray_IsPythonNumber(other)) {
            PyObject *scalar, *result;
            scalar = Py@Name@_From@Name@(arg1);
            result = PyObject_RichCompare(scalar, other, cmp_op);
            if (result == NULL) {
                return NULL;
            }
            out = PyObject_IsTrue(result);
            if (out == -1) {
                return NULL;
            }
            Py_DECREF(result);
            goto done;
        }
#endif
        return PyGenericArrType_Type.tp_richcompare(self, other, cmp_op);
```
But it fails for `np.int64(-1) == 18446744073709551615` because arg1 becomes `-1`. Even if I replace the code with `.item` I get same, anything I can do differently here?

seberg commented:
Ah, right, I am not sure you need the `* #fastpath = 1*14, 0*3#` addition, because longdouble takes the `-3` case?

Why are you converting the `obj` to the scalar instead of the other way around? Your old code with `item = gentype_generic_method(self, NULL, NULL, "item");` made more sense to me (albeit it should be replaced with the correct C-side incarnation).

It all looks weird, but when `other` is an arbitrary Python object, and the numpy scalar is not longdouble/clongdouble, we actually currently already do:

1. Call ufuncs
2. Convert both into an array
    * The second object will be an `object` dtype array
3. Cast the original scalar (now an array) to `object` so that it has the same type as the other
    * This cast, for the main builtin numerical types, actually does `arr.item()` and not just `arr[()]`.

So in the end we already operate on Python objects, just in a very slow and convoluted way.

ganesh-k13 commented:
Oh I see, yeah I missed the `_convert2_to_ctypes` logic. `gentype_generic_method` was static to the file, hence needed conversion to a scalar to call its richcompare, hence the obj to scalar. 

Given that the original commit is showing significant gains in performance, I'll dig in a bit to see where we are slowing down. 

> albeit it should be replaced with the correct C-side incarnation

Yeah, probably moving previous code up a function call should do it, will do so after reading a bit.

seberg commented:
Ohh, I think you are testing mixed types? I.e. one of them is an arbitrary type, the other is always a default integer (probably `C` long?)...  I.e. `int8 + int64` is *super* slow, but `int64 + int8` is fine.

IIRC those paths were messed up, because we did not have a correct "hierarchy" in the code. I tried to clean it up, but it turned out a pretty big PR and I never opened it, I should probably reactivate. There was a problem in that what we should be doing is:
```
if other.dtype.type_num > self.dtype.type_num:
    return NotImplemented
```
(At least that seemed like the easiest solution).  Right now I think what happens it that if the first dtype is the smaller type, we will always end up in the super-slow path. With the `return NotImplemented` fix, we at least only end up in the super-slow path if `(a + b).dtype is not in (a.dtype, b.dtype)`, which is much better.

More complex or generic schemes seemed like over-engineering to me, as they would have to duplicate a lot of ufunc logic which is likely not going to end up very light-weight in any case.

I would love if if you can find a more minimal fix, but in many ways I think this code requires an overhaul :(. Which means rewriting a bunch of it.

ganesh-k13 commented:
I started with your [branch](https://github.com/seberg/numpy/tree/scalar-binop-giveup-in-fallback) and resolved the conflicts(not logical conflicts, as it does not compile yet :) ) here: https://github.com/ganesh-k13/numpy/pull/1. 

I'll be happy to resume work on it and finish it if you feel it helps this part of the code.

seberg commented:
I feel it would help a lot, but I don't quite remember the state :), this code hasn't been touched in more than a decade, at least not seriously. The code is super convoluted, and I at least think it made it quite a bit more easy to grasp the logic (and probably actually fixes a few bugs).

ganesh-k13 commented:
Ohh I see. Cool yeah, I'll work on that branch. Thanks.

ganesh-k13 committed:
MAINT: Refactoring

ganesh-k13 committed:
MAINT: Renamed _typenum_fromtypeobj to PyArray_TypeNumFromNumPyScalarType

ganesh-k13 committed:
ENH: Added descr_from_basic_scalar

ganesh-k13 committed:
ENH: Added fast path for richcompare

ganesh-k13 committed:
ENH: Added comments and header

ganesh-k13 commented:
Hey @seberg , I leveraged some stuff from the branch and added fastpath in the `-1` case where it's safe to cast. Let me know your thoughts on this.

ganesh-k13 committed:
ENH: Replaced `item` getting from array to getitem call

ganesh-k13 commented:
```
       before           after         ratio
     [4d290795]       [3115c165]
     <master>         <enh_14415_fast_compare>
-      23.7±0.1μs      11.7±0.04μs     0.49  bench_scalar.ScalarMath.time_compare('int32')
-      23.9±0.3μs       11.7±0.2μs     0.49  bench_scalar.ScalarMath.time_compare('int16')
-      25.5±0.3μs       12.4±0.2μs     0.49  bench_scalar.ScalarMath.time_compare('complex64')
-     25.5±0.05μs      11.9±0.09μs     0.47  bench_scalar.ScalarMath.time_compare('float32')
-      26.1±0.4μs      11.9±0.05μs     0.45  bench_scalar.ScalarMath.time_compare('float16')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

```

ganesh-k13 committed:
BENCH: Moved benchmarks for fast paths (#17970)

ganesh-k13 committed:
BUG: Moved Py_DECREF out of block | Init for descr

BUG: handled ret and overflow

BUG: Fixed crash in do_richcompare_on_scalars | Added complex check on fastpath

BUG: Added scalar check for self

MAINT: Refactor PyLong descr creation

MAINT: Moved header location

MAINT: Added XDECREF | Refactor complex equality check

BUG: Fixed error in complex check logic

ganesh-k13 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -569,6 +570,88 @@ static void
+                return -1;
+            }
+            /* Then fail if there is an overflow(-1 or 1) */
+            assert(overflow == 0);
```
This should return `-1` along with assert, will fix this

ganesh-k13 commented:
Took some time to repro the issue, we hit a NULL pointer dereference(Py_DECREF on NULL) in complex number cases. Will fix it.

ganesh-k13 commented:
Hi @seberg , sorry for the long pause in the PR.
I noticed the following TC causing crash in specific environment. I think handling return value for `self_descr` for overflow should take care of it. 
Now to the overflow: https://github.com/numpy/numpy/blob/a190258d4e90f2a17a9469e5dd9fb5f4b045aa90/numpy/core/tests/test_getlimits.py#L68-L70
Is it defined behaviour to do `np.uint(-1)`?

charris commented:
> Is it defined behaviour to do np.uint(-1)?

I think so, although it may not be formalized. C allows it, but maybe we should make an explicit choice.

seberg commented:
Lets say that it was once explicitly implemented that `uint(-1)` will work, so for now I will say yes. For comparisons it would be best if this can't happen (the only problematic thing is probably uint64 == int64 using float and losing precision though).

It is something that we could think about changing, but I am not sure it is worth the trouble and how annoying the work around is (since using -1 for maxint is not uncommon, even if technicall undefined by the C standard.)

ganesh-k13 commented:
Hey @seberg , I tried multiple things to reproduce the issue locally, tried `quay.io/pypa/manylinux2010_i686` and used a native 32-bit system as well, but it's not crashing. Any idea why that particular TC is failing? I followed the memory through the code, no double free or invalid dereference either.

ganesh-k13 commented:
Also not sure if I am setting `quay.io/pypa/manylinux2010_i686` correctly, for instance, I don't see python being installed in the image, yet it's there. I was curious about how are the packages being installed in the image.

seberg commented:
Maybe that overflow `assert` is simply incorrect? It does seem a bit off.  Otherwise, my best idea is to install `gdb` in the CI and the job with that. Hopefully, getting a traceback will be sufficient. I did something for example [here (the command may be useful)](https://github.com/numpy/numpy/pull/17750/files) (but for the travis CI).

ganesh-k13 commented:
Thanks @seberg , I'll try to refactor that bit. I had asked the python guys here: https://bugs.python.org/issue42665 , seems like we can use PyLong_AsLong directly and try. The gdb in CI is very useful, thanks.

ganesh-k13 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -569,6 +570,91 @@ static void
+        *descr = PyArray_DescrFromType(type_num);
+        return 0;
+    }
+    else if (PyFloat_CheckExact(obj)) {
```
We could add an error check here as well?

seberg replied:
Sorry, where do you mean? The type check and `DescrFromType` cannot fail here (the latter is a bit obscure sometimes, I admit, but still cannot fail since it returns a singleton).

ganesh-k13 replied:
```suggestion
    else if (PyFloat_CheckExact(obj)) {
        if (PyFloat_AsDouble(obj) < 0 && PyErr_Occurred()) {
            PyErr_Clear();
            return -1;
        }
```
So I mean like this, to make sure the object can be interpreted safely as a `float`

seberg replied:
`PyFloat_AsDouble` cannot possibly fail after the `PyFloat_CheckExact`, at least in cpython. So I don't think we have to add it.

ganesh-k13 replied:
Ohh thanks, I did not know that.

ganesh-k13 commented:
Benchmarks after the changes:
```
       before           after         ratio
     [f5f845b7]       [fe3a8975]
     <master>         <enh_14415_fast_compare>
+         583±4ns          622±4ns     1.07  bench_scalar.ScalarMath.time_addition('longfloat')
-     3.63±0.03μs      3.45±0.03μs     0.95  bench_scalar.ScalarMath.time_addition_pyint('float32')
-       584±0.8ns          554±2ns     0.95  bench_scalar.ScalarMath.time_multiplication('complex256')
-         563±7ns          531±2ns     0.94  bench_scalar.ScalarMath.time_addition('float32')
-         567±2ns          527±3ns     0.93  bench_scalar.ScalarMath.time_multiplication('int64')
-      23.7±0.2μs      11.5±0.07μs     0.49  bench_scalar.ScalarMath.time_compare('int32')
-      24.1±0.7μs       11.7±0.2μs     0.48  bench_scalar.ScalarMath.time_compare('int16')
-      25.9±0.3μs      12.1±0.05μs     0.47  bench_scalar.ScalarMath.time_compare('complex64')
-      26.2±0.2μs      12.0±0.07μs     0.46  bench_scalar.ScalarMath.time_compare('float32')
-      26.3±0.5μs      11.9±0.03μs     0.45  bench_scalar.ScalarMath.time_compare('float16')
```

ganesh-k13 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -15,6 +15,7 @@
 #include "numpy/arrayobject.h"
 #include "numpy/ufuncobject.h"
 #include "numpy/arrayscalars.h"
+#include "scalartypes.h"
```
Is this write place to add it? Any logical grouping of the headers done here? I noticed there are groups

seberg replied:
Honestly, the header includes are probably just a mess all over the place, and I am not sure of much logic... I would probably group it by where the folders (which may mean putting it by itself here). I doubt the rest follows any such logic though. I am willing to bet that most of the order is simply the historical order for when it was necessary to add them...

ganesh-k13 replied:
Oh I see. Yeah, I'll make it separate.

ganesh-k13 commented:
Hey @seberg , can you have a look at this PR. Thanks.

seberg commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -569,6 +571,91 @@ static void
+/**
+ * This function attempts to compare NumPy or Python scalars
+ * The operation is done by getting value of both scalars
+ * and calling richcompare on them.
```
Technically, right now it only gets the value of one scalar, and then does a recursive call, right now? Maybe we should even just get the value of both scalars though and do it as it is documented.

ganesh-k13 replied:
It got super interesting when I tried this :) :
```
-            if (item_self != NULL) {
-                ret = PyObject_RichCompare(item_self, other, cmp_op);
+            data_other = scalar_value(other, NULL);
+            item_other = other_descr->f->getitem(data_other, NULL);
+            if (item_self != NULL && item_other != NULL) {
+                ret = PyObject_RichCompare(item_self, item_other, cmp_op);
             }
             Py_XDECREF(item_self);
+            Py_XDECREF(item_other);
```

Here is output:
```
>>> np.float16(3) == 3
False
>>> np.int16(3) == 3
False
>>> np.int32(3) == 3
False
>>> np.int64(3) == 3
True
>>> np.int64(3) == 3
True
```

I shall get to the bottom of this

ganesh-k13 replied:
Ok I'll circle back to this after the tests, I reached still here and got confused more as to what's happening: https://github.com/python/cpython/blob/c9b8e9c421b57acdcaf24fab0c93bc29b3ef7c67/Objects/object.c#L658

seberg commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -569,6 +571,91 @@ static void
+            (other_descr != NULL && other_descr->type_num < NPY_CFLOAT))
+            || (cmp_op == Py_EQ || cmp_op == Py_NE)) {
+            data_self = scalar_value(self, NULL);
+            item_self = self_descr->f->getitem(data_self, NULL);
```
This can fail (return `NULL`), although RichCompare may guard against it, we should probably just check it here.

seberg reviewed (commented):
Thanks, it seems you got it figured out. Pretty much just style nits.

I think we need a test with `np.float128` as both first and second argument (and maybe both), just to double check we are not going into the recursion trap.  The other thing to test might be `np.float16`, just because it is a bit odd (e.g. its typenumber is higher than the other numbers, a fact I conveniently forgot for a while).

ganesh-k13 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -569,6 +571,91 @@ static void
+        /* If either of the operands are complex AND the operator is not equality,
+         * python's built-in richcompare cannot be used as it is not supported */
+        if ((self_descr->type_num < NPY_CFLOAT &&
+            (other_descr != NULL && other_descr->type_num < NPY_CFLOAT))
```
I attempted to simplify that `if`, please do let me know if that's ok

ganesh-k13 commented:
Thanks for the review @seberg , I shall fix the scalar conversion of `other` and then add the tests you mentioned as well.

ganesh-k13 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -569,6 +571,92 @@ static void
+
+    if (pyscalar_self >= 0 && pyscalar_other >= 0) {
+        /*
+         * If either of the operands are complex AND the operator is not equality,
```
```suggestion
         * If both operands are complex AND the operator is not equality,
```
meant this, will fix

ganesh-k13 replied:
Ok this comment is actually right the logic is incorrect :) , will fix it.

seberg replied:
Thanks! Just make sure we have a test, that would be great ;)  (I admit it can be pretty annoying in this type of context where the tests exist a bit, but not thorough.)

ganesh-k13 replied:
Luckily I found this via a test(https://github.com/numpy/numpy/pull/17970/checks?check_run_id=1767648779#step:4:5730), but yeah, I'll see how to make it a bit more thorough

ganesh-k13 commented:
Hey @seberg sorry for the long pause. Basically when I added `np.float128` test i noticed a crash. I'm working on fixing that adding some more cases.

seberg commented:
Long pauses are typical, just make sure to bug me/us when you feel it would help. It is pretty heroic to dabble in these parts of NumPy!

ganesh-k13 commented:
Thanks a lot @seberg :). The issue is a bit strange actually.

Say I run a this piece of snippet:
```py
f16 = np.float128(); b = bool(); b_ = np.bool_()
f16 == b
f16 == b_
```
in this [line](https://github.com/numpy/numpy/pull/17970/files#diff-7355537178d1b6f85a40acf970ac3317b184721c75828f33ed136ea0f6c92ae9R647) in case of `bool` this happens:
```
Breakpoint 1, LONGDOUBLE_getitem (ip=0xf73f71a0, ap=0x7ffff71ad3c3 <is_anyscalar_exact+32>) at numpy/core/src/multiarray/arraytypes.c.src:430
430     {
(gdb) f 1
#1  0x00007ffff730c511 in do_richcompare_on_scalars (self=0x7ffff5ed1e70, other=0x7ffff73f71a0 <_PyArrayScalar_BoolValues>, cmp_op=2) at numpy/core/src/umath/scalarmath.c.src:647
647                 item_self = self_descr->f->getitem(data_self, NULL);
```

Although we call getitem with NULL, ap has some value. Like how is that even possible :) 
but for `np.bool_`:
```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff704fce8 in PyArray_DESCR (arr=0x0) at numpy/core/include/numpy/ndarraytypes.h:1536
1536        return ((PyArrayObject_fields *)arr)->descr;
(gdb) f 1
#1  0x00007ffff705377b in LONGDOUBLE_getitem (ip=0x7ffff5ed1e80, ap=0x0) at numpy/core/src/multiarray/arraytypes.c.src:431
431         return PyArray_Scalar(ip, PyArray_DESCR((PyArrayObject *)ap), NULL);
```

Naturally we hit a null pointer dereference and crash. What is super weird is that off all the combinations of all dataypes in numpy and python(I tried [this](https://gist.github.com/ganesh-k13/6201227c5d3d65902c6eaf71357e72b1#file-comparision-py)), only a `float128` with `np.bool_` gives this error and that line has nothing to do with the other operand.

Any pointers on this would be very helpful.

seberg commented:
Umm, one thing is that most of *our* `getitem` functions do have this line `if ((ap == NULL) || PyArray_ISBEHAVED_RO(ap)) {` and I think there is some internal code that assumes passing `NULL` will work for basic/naive dtypes.

Honestly, while calling with `NULL` should not be done (especially outside of NumPy), I think this is "by design" that we do call it with `NULL` for basic dtypes in a few places. This happens to be one of them, and I guess before that path was never taken for `longdouble`.

ganesh-k13 commented:
Ohh I see, yeah in other places, we are not calling getitem in a way we are calling here. I am experimenting with a few NumPy c-api for getting the scalar values instead of just calling getitem. I think one of them will work eventually, I'll keep you updated. Thanks.

ganesh-k13 commented:
I asked the python guys and seems like `long double` is not natively supported. I guess I'll treat it as a non-native type itself and skip it from native richcompare.
Ref: https://bugs.python.org/issue43177

seberg commented:
Sorry about that, I did not realize that was confusing. Yes, longdouble cannot be converted to a Python type (or clongdouble), since Python only has `float64`.  But `longdouble` is also not worth all that much headaches probably, lets focus on the common types instead, its OK if longdouble keeps using the "super slow paths".

ganesh-k13 committed:
ENH: Enhanced scalar fastpaths

ganesh-k13 committed:
BENCH: Added scalar-scalar checks (#17970)

ganesh-k13 commented:
Updated bench for changes:
```
       before           after         ratio
     [f6a71fea]       [13e94849]
     <master>         <enh_14415_fast_compare>
+         455±2ns         545±60ns     1.20  bench_scalar.ScalarMath.time_abs('longfloat')
+         490±7ns        585±100ns     1.19  bench_scalar.ScalarMath.time_addition('int64')
+         539±6ns         579±20ns     1.07  bench_scalar.ScalarMath.time_addition('complex256')
+       538±0.6ns         571±20ns     1.06  bench_scalar.ScalarMath.time_addition('longfloat')
+         485±1ns          513±6ns     1.06  bench_scalar.ScalarMath.time_multiplication('float64')
+         483±5ns         511±20ns     1.06  bench_scalar.ScalarMath.time_addition('float64')
+       456±0.7ns          482±3ns     1.06  bench_scalar.ScalarMath.time_abs('complex128')
+         489±6ns          515±5ns     1.05  bench_scalar.ScalarMath.time_multiplication('int64')
-      82.9±0.6μs       54.8±0.5μs     0.66  bench_scalar.ScalarMath.time_compare_types('float64')
-         128±4μs         75.4±2μs     0.59  bench_scalar.ScalarMath.time_compare_types('int64')
-         152±7μs         76.3±2μs     0.50  bench_scalar.ScalarMath.time_compare_types('int32')
-         117±8μs       58.7±0.2μs     0.50  bench_scalar.ScalarMath.time_compare_types('complex64')
-         161±4μs         77.5±1μs     0.48  bench_scalar.ScalarMath.time_compare_types('int16')
-         129±2μs       59.1±0.4μs     0.46  bench_scalar.ScalarMath.time_compare_types('float32')
-     19.6±0.05μs       8.61±0.1μs     0.44  bench_scalar.ScalarMath.time_compare('int32')
-      20.0±0.4μs       8.36±0.2μs     0.42  bench_scalar.ScalarMath.time_compare('int16')
-      21.7±0.5μs      8.94±0.03μs     0.41  bench_scalar.ScalarMath.time_compare('complex64')
-      21.8±0.1μs       8.73±0.2μs     0.40  bench_scalar.ScalarMath.time_compare('float32')

```

ganesh-k13 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -569,6 +571,112 @@ static void
+        is_complex_operands = (PyTypeNum_ISCOMPLEX(self_descr->type_num) ||
+            PyTypeNum_ISCOMPLEX(other_descr->type_num));
+        is_equality_operator = (cmp_op != Py_EQ && cmp_op != Py_NE);
+        is_python_scalar = PyDescr_PythonCRepresentable(self_descr) ||
```
I think a less confusing name would be `is_not_*` and `PyDescr_Not*`, will fix, aim is to basically check if the value can be represented as a native c type that python allows.

seberg replied:
No need for the "C" probably, we are aiming for converting (safely) to a native Python type?

ganesh-k13 replied:
Yes indeed, also the `PyArray_Is*` functions that are already present do not have a proper check for native python types.

ganesh-k13 replied:
The problem being that, if we cannot represent in a PyObject, the richcompare seems to cause problem, long double being one of the victims.

ganesh-k13 commented:
Also I found a SIGABORT in the bench where I added a check for a type compared with all other dtype [here](https://github.com/numpy/numpy/pull/17970/files#diff-62ea5eaf0f2bbf20a3cd8fff578ceb0b32391a162a2f7896f5977af6029689a7R42). But no crash in UT. In particualr it means we lack a test where a `clongdouble` is never compared with `int`. I feel we must add this test somewhere. Any good location to add this?

ganesh-k13 committed:
MAINT: Added comments to richcompare fastpath

ganesh-k13 committed:
TST: Added numbers comparison (#17970)

ganesh-k13 commented:
Hey @seberg, let me know any other changes needed here. Thanks.

ganesh-k13 commented:
I think the main pending items from side are the tests, overflow and few other combinations, will add them.

ganesh-k13 commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -569,6 +571,91 @@ static void
+    else if (PyLong_CheckExact(obj)) {
+        if (PyLong_AsLong(obj) == -1 && PyErr_Occurred()) {
+            PyErr_Clear();
+            return -1;
```
Hey @seberg , any idea how we can hit this reliably? We don't reach case 2 in case of raw int, at the same time, we cannot overflow a Numpy scalar while creating it.

seberg commented:
Thanks @ganesh-k13, I will prioritize this a bit soon. gh-18548 also makes this a bit timely, although it is probably somewhat orthogonal it struggles with similar problems...

seberg commented:
For myself, maybe it does not matter: But if I get it right, doing this would basically cement the behaviour in gh-10322 for scalars.  I really hope we can get away with fixing gh-10322 though, because if we can't the whole "weak python scalar" idea might break down. But if it breaks down, I am left with shambles of confusion because there is no likely chance of a consistent future :(.

seberg commented:
Going to close, since the merge conflicts will be large now.  There are still things that this tries to speed up that are not sped up in the meantime (e.g. `1j == np.float64(1)`), but a few other paths are improved anyway.
Hope you don't mind, if you do just reopen :).
